### PR TITLE
core: make StateDirectory= or friends support '-' prefix

### LINKED ISSUE #9296: RFE: support '-' prefix for StateDirectory= or friends
systemd-timesyncd cannot be used on a read only file system. I have attempted changing the symlink from ../private/systemd/timesync/ to /run/timesync (on a tmpfs) and restarting the service and it complains every time even though ownership is correct. I can run the program from the command line and it appears to use the new location on /run/timesync for the 'clock' file as the creation time does change when I run it.  I have been unable to get it to start and run as a service, and use /run, though.

I am attempting to set up my raspberry pi with a readonly file system and am struggling with this nugget. 

What would be nice is to pass an argument to systemd-timesyncd to tell it where to write its 'clock' file. 

"/usr/lib/systemd/systemd-timesyncd -p /run/timesyncd"

If there is a way to achieve this by way of a configuration of the systemd-timesyncd.service file, I'd love to know about it.

### PR DESCRIPTION
**Pull request by yuwata**
If a path has '-' prefix, then any errors in preparing the directory will be removed.

Closes #9296.

### CODE REVIEW

**Comment by yuwata** [comment]
Note. This breaks a backward compatibility. If a unit has `StateDirectory=-foo`, then before this, /var/lib/-foo was created, but after this, /var/lib/foo will be created. Please give me any comments and suggestions.

File: src/core/dbus-execute.c [hunk 1]
```diff
@@ -2087,52 +2124,53 @@ int bus_exec_context_set_transient_property(
...
+                assert(i >= 0 && i < _EXEC_DIRECTORY_TYPE_MAX);
+
                 r = sd_bus_message_read_strv(message, &l);
                 if (r < 0)
                         return r;
 
                 STRV_FOREACH(p, l) {
-                        if (!path_is_normalized(*p))
-                                return sd_bus_error_setf(error, SD_BUS_ERROR_INVALID_ARGS, "%s= path is not normalized: %s", name, *p);
+                        char *q = *p + (**p == '-');
```
Comment (keszybz) [inline src/core/dbus-execute.c]: `const char *` ...

File: src/core/dbus-execute.c [hunk 2]
```diff
@@ -2087,52 +2124,53 @@ int bus_exec_context_set_transient_property(
...
                         if (strv_isempty(l)) {
-                                *dirs = strv_free(*dirs);
+                                exec_directories_clear(ed);
                                 unit_write_settingf(u, flags, name, "%s=", name);
-                        } else {
-                                _cleanup_free_ char *joined = NULL;
+                        } else
+                                STRV_FOREACH(p, l) {
+                                        bool ignore = false;
+                                        char *q = *p;
```
Comment (keszybz) [inline src/core/dbus-execute.c:2154]: `const` here too?

File: src/core/execute.c [hunk 3]
```diff
@@ -3963,10 +4038,15 @@ void exec_context_dump(const ExecContext *c, FILE* f, const char *prefix) {
...
+
                 fprintf(f, "%s%sMode: %04o\n", prefix, exec_directory_type_to_string(dt), c->directories[dt].mode);
 
-                STRV_FOREACH(d, c->directories[dt].paths)
-                        fprintf(f, "%s%s: %s\n", prefix, exec_directory_type_to_string(dt), *d);
+                for (j = 0; j < c->directories[dt].n_directories; j++)
+                        fprintf(f, "%s%s%s: %s\n", prefix,
+                                exec_directory_type_to_string(dt),
+                                c->directories[dt].directories[j].ignore ? "-" : "",
+                                c->directories[dt].directories[j].path);
```
Comment (keszybz) [inline src/core/execute.c]: This gives something like "RuntimeDirectory-: foobar" (the dash is in the wrong place).

**Comment by keszybz** [comment]
> If a unit has StateDirectory=-foo, then before this, /var/lib/-foo was created, but after this, /var/lib/foo will be created.

I think this is OK. Let's assume that nobody creates directories like that. (Even though we have `-.slice`, yuck.)

File: src/core/dbus-execute.c (hunk 2 above)
Comment (yuwata) [reply src/core/dbus-execute.c:2154]: It is possible, but `.path` is `char *`. Casting from `char*` to `const char*` here and revert it at the later substitution may be not good, I think.

File: src/core/dbus-execute.c (hunk 1 above)
Comment (yuwata) [reply src/core/dbus-execute.c]: It is not necessary, but ok.

File: src/core/execute.c (hunk 3 above)
Comment (yuwata) [reply src/core/execute.c]: ouch. fixed.

**Comment by yuwata** [comment]
Updated. I thought that this may be useful not only for the case #9296, but also `RootImage/RootDirectory=` with read-only image or directory. But it may need more work to support such read-only systems. I will test this more in detail. So, tentatively, I have added 'dont-merge' tag.

**Comment by yuwata** [comment]
Closing this, as it does not work at least in the current form.

**Pull request closed without merging**